### PR TITLE
ARROW-2638: [Python] Prevent calling extension class constructors directly

### PR DESCRIPTION
**Pull request by kszucs**
Using http://docs.cython.org/en/latest/src/userguide/extension_types.html#fast-instantiation

What do You think @pitrou ? Should I implement for the rest of the classes? (Field, Schema etc.)

### CODE REVIEW

File: python/pyarrow/table.pxi [hunk 1]
```diff
@@ -565,6 +565,10 @@ cdef class RecordBatch:
         self.batch = NULL
         self._schema = None
 
+    def __init__(self):
+        raise RuntimeError("Do not call RecordBatch's constructor directly, "
+                           "use one of the `from_*` methods instead.")
```
Comment (pitrou) [inline python/pyarrow/table.pxi]: Nit: they are not methods but functions. Or perhaps there are both, so say "methods and functions".

File: python/pyarrow/tests/test_schema.py [hunk 2]
```diff
@@ -23,6 +23,13 @@
 import pyarrow as pa
 
 
+def test_schema_constructor_errors():
+    with pytest.raises(ReferenceError):
+        repr(pa.Schema())
```
Comment (pitrou) [inline python/pyarrow/tests/test_schema.py]: I know you are just moving those tests around, but I wonder why `repr()` needs to be called here. Shouldn't the constructor itself raise an error?

File: python/pyarrow/table.pxi [hunk 3]
```diff
@@ -565,6 +565,10 @@ cdef class RecordBatch:
         self.batch = NULL
         self._schema = None
 
+    def __init__(self):
+        raise RuntimeError("Do not call RecordBatch's constructor directly, "
```
Comment (pitrou) [inline python/pyarrow/table.pxi]: I would raise `TypeError`, not `RuntimeError`.

**Review by pitrou — COMMENTED** [review]
This looks ok on the principle. I think this can be extended to all other such classes.

File: python/pyarrow/tests/test_schema.py (hunk 2 above)
Comment (kszucs) [reply python/pyarrow/tests/test_schema.py]: I'm just writing for Schema and DataTypes, it'll raise TypeError too

File: python/pyarrow/table.pxi (hunk 3 above)
Comment (kszucs) [reply python/pyarrow/table.pxi]: Ok

**Comment by pitrou** [comment]
Using `__reduce__` is fine with me. We may be breaking binary pickle compatibility with previous versions, but at this point I think it's not a problem.

**Comment by xhochy** [comment]
> Using __reduce__ is fine with me. We may be breaking binary pickle compatibility with previous versions, but at this point I think it's not a problem.

I would not take this at the moment as a concern. Pickling in old releases was quite broken. Master is behaving much better now.

**Comment by kszucs** [comment]
Are the `_check_nullptr` methods still required? It looks like the only possible way the class can be in uninitialized state was calling the constructors, but that's prevented now.

File: python/pyarrow/tests/test_types.py [hunk 4]
```diff
@@ -53,6 +53,31 @@
...
+           "instead.".format(cls=klass.__name__, func=klass.__name__.lower()))
+
+    with pytest.raises(TypeError, match=msg):
+        klass()
+
+
+@pytest.mark.parametrize('klass', [
+    pa.DataType,
+    pa.TimestampType,
+    # others are not part of the public API
```
Comment (kszucs) [inline python/pyarrow/tests/test_types.py]: Is this intentional? It looks assymetric.

**Comment by pitrou** [comment]
> Are the _check_nullptr methods still required? 

I don't think so.

File: python/pyarrow/tests/test_types.py (hunk 4 above)
Comment (pitrou) [reply python/pyarrow/tests/test_types.py]: I don't know. If it's not too much work you could test all of them.

Comment (kszucs) [reply python/pyarrow/tests/test_types.py]: I mean others are not exposed. Should I expose them?

Comment (pitrou) [reply python/pyarrow/tests/test_types.py]: Who do you mean with "expose" exactly?

Comment (kszucs) [reply python/pyarrow/tests/test_types.py]: I'm testing under `pa.lib.*Type` instead of `pa.*Type`.

**Comment by kszucs** [comment]
@pitrou Is that Ok if I try to remove `_check_nullptr` methods in a subsequent PR?

File: python/pyarrow/tests/test_types.py (hunk 4 above)
Comment (pitrou) [reply python/pyarrow/tests/test_types.py]: Ah, I see. I wonder what the rationale is for exposing `TimestampType`. @xhochy

**Comment by pitrou** [comment]
> Is that Ok if I try to remove _check_nullptr methods in a subsequent PR?

That's fine with me :-)

**Comment by kszucs** [comment]
@pitrou Do We want to allow subclassing extension types?

**Comment by pitrou** [comment]
> Do We want to allow subclassing extension types?

I don't think we care either way. Let users subclass them, but without any guarantee.

File: python/pyarrow/tests/test_types.py (hunk 4 above)
Comment (xhochy) [reply python/pyarrow/tests/test_types.py]: I cannot remember why. @wesm any ideas? I think we only use `pa.timestamp()` at the moment from an end-user perspective.

**Comment by kszucs** [comment]
I've removed `DataType`, `NAType` and `TimestampType` from public namespace

**Comment by kszucs** [comment]
Also resolves https://issues.apache.org/jira/projects/ARROW/issues/ARROW-2639

**Comment by kszucs** [comment]
Affected classes:
```python
pa.Field,
pa.Schema,
pa.Column,
pa.ChunkedArray,
pa.RecordBatch,
pa.Table,
pa.Buffer,
pa.Array,
pa.Tensor,
pa.lib.DataType,
pa.lib.ListType,
pa.lib.UnionType,
pa.lib.StructType,
pa.lib.Time32Type,
pa.lib.Time64Type,
pa.lib.TimestampType,
pa.lib.Decimal128Type,
pa.lib.DictionaryType,
pa.lib.FixedSizeBinaryType
```

**Comment by kszucs** [comment]
@pitrou PlasmaBuffer breaks the pattern, how should we handle that?

**Comment by kszucs** [comment]
@wesm waiting for review

File: python/pyarrow/__init__.py [hunk 5]
```diff
@@ -53,7 +53,6 @@ def parse_version(root):
                          binary, string, decimal128,
                          list_, struct, union, dictionary, field,
                          type_for_alias,
-                         DataType, NAType,
```
Comment (wesm) [inline python/pyarrow/__init__.py]: I think it'd be useful to have `pyarrow.DataType` still

**Review by wesm — APPROVED** [review]
Looks good except for the `DataType` thing. I'll see if I can restore that quickly and then we can merge

**Comment by wesm** [comment]
Thanks for doing this cleanup; much better than what we were doing before -- I wasn't aware of the `__new__` trick to circumvent `__init__` with Cython extension types

**Comment by kszucs** [comment]
Thanks!

**Comment by wesm** [comment]
Rebased

**Pull request closed without merging**